relayer/relays/parachain/scanner.go

```diff
@@ -0,0 +1,544 @@
+package parachain
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"sort"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	log "github.com/sirupsen/logrus"
+	gsrpc "github.com/snowfork/go-substrate-rpc-client/v4"
+	"github.com/snowfork/go-substrate-rpc-client/v4/types"
+	"github.com/snowfork/snowbridge/relayer/chain/ethereum"
+	"github.com/snowfork/snowbridge/relayer/chain/parachain"
+	"github.com/snowfork/snowbridge/relayer/chain/relaychain"
+	"github.com/snowfork/snowbridge/relayer/contracts/basic"
+	"github.com/snowfork/snowbridge/relayer/contracts/incentivized"
+)
+
+type Scanner struct {
+	config           *SourceConfig
+	ethConn          *ethereum.Connection
+	relayConn        *relaychain.Connection
+	paraConn         *parachain.Connection
+	paraID           uint32
+	tasks            chan<- *Task
+	eventQueryClient QueryClient
+	accounts         [][32]byte
+}
+
+// Scans for all parachain message commitments that need to be relayed and can be proven
+// using the MMR root at the specified beefyBlockNumber of the relay chain.
+//
+// The algorithm works roughly like this:
+// 1. Fetch channel nonces on both sides of the bridge and compare them
+// 2. If the nonces on the parachain side are larger that means messages
+//    need to be relayed. If not then exit early.
+// 3. Scan parachain blocks to figure out exactly which commitments need to be relayed,
+// 4. For all the parachain blocks with unsettled commitments, determine the relay chain
+//    block number in which the parachain block was included,
+//
+func (s *Scanner) Scan(ctx context.Context, beefyBlockNumber uint64) ([]*Task, error) {
+	beefyBlockHash, err := s.relayConn.API().RPC.Chain.GetBlockHash(uint64(beefyBlockNumber))
+	if err != nil {
+		return nil, fmt.Errorf("fetch block hash for block %v: %w", beefyBlockNumber, err)
+	}
+
+	// fetch last parachain header that was finalized *before* the BEEFY block
+	beefyBlockMinusOneHash, err := s.relayConn.API().RPC.Chain.GetBlockHash(uint64(beefyBlockNumber-1))
```

Comment:
Why do we look at the second-last beefy block here? Are the parachain headers in the last block not finalized yet?

Reply:
Its more that if a BEEFY commitment was generated for block X, then the MMR root hash in the commitment can only be used to prove parachain headers included before block X.

This explains it a bit more: https://docs.snowbridge.network/architecture/verification/polkadot#message-verification